### ISSUE TITLE
Consider removing reduce and derived methods

### ISSUE THREAD

**Issue by jwoudenberg** [elm-community/lazy-list#9]
The `reduce` operation and everything based on it (`foldl`, `foldr`, `sum`, `product`, perhaps others?) aren't lazy in that they force evaluation of the entire list. Writing this down it feels very obvious, but I remember that while working with some lazy lists in the `Fuzzer` codebase it took me some time to realise a `foldl` in the code was creating a performance problem.

If you're using a lazy list then evaluating the entire list is probably not what you want. If you do really want it, it's easy even without using the functions this library provides: Simply turn your lazy list into a standard list, then call the relevant function from the `List` module. This approach clearly signposts that we're leaving lazy territory by the call to `toList`.

So in summary, I think removing these functions prevents some easy mistakes with a minimal impact on usability. Looking forward to your thoughts!

**Comment by GordonBGood** [comment]
There is just one objection to your proposal that I can think of: pre-evaluating a lazy list into a list means that the resulting list will simultaneously occupy all the memory required for every element of the resulting list, whereas properly evaluating the lazy list element-by-element as it is consumed means there is never more memory needed than that required for a single element at a time.
